include/deal.II/lac/solver_idr.h

```diff
@@ -0,0 +1,475 @@
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverIDR() override = default;
+
+  /**
+   * Solve the linear system <code>Ax=b</code> for x.
+   */
+  template <typename MatrixType, typename PreconditionerType>
+  void
+  solve(const MatrixType &        A,
+        VectorType &              x,
+        const VectorType &        b,
+        const PreconditionerType &preconditioner);
+
+protected:
+  /**
+   * Interface for derived class. This function gets the current iteration
+   * vector, the residual and the update vector in each step. It can be used
+   * for graphical output of the convergence history.
+   */
+  virtual void
+  print_vectors(const unsigned int step,
+                const VectorType & x,
+                const VectorType & r,
+                const VectorType & d) const;
+
+private:
+  /**
+   * Additional solver parameters.
+   */
+  AdditionalData additional_data;
+};
+
+/*@}*/
+/*------------------------- Implementation ----------------------------*/
+
+#ifndef DOXYGEN
+
+
+namespace internal
+{
+  namespace SolverIDRImplementation
+  {
+    template <class VectorType>
+    inline TmpVectors<VectorType>::TmpVectors(const unsigned int        s_param,
+                                              VectorMemory<VectorType> &vmem)
+      : mem(vmem)
+      , data(s_param)
+    {}
+
+
+
+    template <class VectorType>
+    inline VectorType &TmpVectors<VectorType>::
+                       operator[](const unsigned int i) const
+    {
+      Assert(i < data.size(), ExcIndexRange(i, 0, data.size()));
+
+      Assert(data[i] != nullptr, ExcNotInitialized());
+      return *data[i];
+    }
+
+
+
+    template <class VectorType>
+    inline VectorType &
+    TmpVectors<VectorType>::operator()(const unsigned int i,
+                                       const VectorType & temp)
+    {
+      Assert(i < data.size(), ExcIndexRange(i, 0, data.size()));
+      if (data[i] == nullptr)
+        {
+          data[i] = std::move(typename VectorMemory<VectorType>::Pointer(mem));
+          data[i]->reinit(temp);
+        }
+      return *data[i];
+    }
+  } // namespace SolverIDRImplementation
+} // namespace internal
+
+
+
+template <class VectorType>
+SolverIDR<VectorType>::SolverIDR(SolverControl &           cn,
+                                 VectorMemory<VectorType> &mem,
+                                 const AdditionalData &    data)
+  : SolverBase<VectorType>(cn, mem)
+  , additional_data(data)
+
+{}
+
+
+
+template <class VectorType>
+SolverIDR<VectorType>::SolverIDR(SolverControl &cn, const AdditionalData &data)
+  : SolverBase<VectorType>(cn)
+  , additional_data(data)
+
+{}
+
+
+
+template <class VectorType>
+void
+SolverIDR<VectorType>::print_vectors(const unsigned int,
+                                     const VectorType &,
+                                     const VectorType &,
+                                     const VectorType &) const
+{}
```

Comment:
Any reason not to implement this? Won't `VectorType::print(std::cout);` work for all `VectorType`s?

Reply:
I copied this function from `solver_cg.h`. In every `solver_*.h` file (except `solver_gmres.h` where there is no `print_vectors()` function) this function is implemented to do nothing. I never call this function so I will just remove it.

Reply:
This function is there to be able to print intermediate solutions by using `DoFPrintSolverStep` with a solver class as base class. It might not be a bad idea to just call this function after updating the solution vector.

Separately, we also need tests for `DoFPrintSolverStep` anyway as there a none...

Reply:
Ok, I'll keep it as is and add a `print_vectors()` call after each update like in `solver_cg.h`.